src/man-pages/man/man3/fnmatch.3
'\" t
.\" Copyright, the authors of the Linux man-pages project
.\"
.\" SPDX-License-Identifier: Linux-man-pages-copyleft
.\"
.TH fnmatch 3 2025-09-21 "Linux man-pages (unreleased)"
.SH NAME
fnmatch \- match filename or pathname
.SH LIBRARY
Standard C library
.RI ( libc ,\~ \-lc )
.SH SYNOPSIS
.nf
.B #include <fnmatch.h>
.P
.BI "int fnmatch(const char *" "pattern" ", const char *" string ", int " flags );
.fi
.SH DESCRIPTION
The
.BR fnmatch ()
function checks whether the
.I string
argument matches the
.I pattern
argument, which is a shell wildcard pattern (see
.BR glob (7)).
.P
The
.I flags
argument modifies the behavior;
it is the bitwise OR of zero or more
of the following flags:
.TP
.B FNM_NOESCAPE
If this flag is set, treat backslash as an ordinary character,
instead of an escape character.
.TP
.B FNM_PATHNAME
If this flag is set, match a slash in
.I string
only with a slash in
.I pattern
and not by an asterisk (*) or a question mark (?) metacharacter,
nor by a bracket expression ([]) containing a slash.
.TP
.B FNM_PERIOD
If this flag is set, a leading period in
.I string
has to be matched exactly by a period in
.IR pattern .
A period is considered to be leading if it is the first character in
.IR string ,
or if both
.B FNM_PATHNAME
is set and the period immediately follows a slash.
.TP
.B FNM_FILE_NAME
This is a GNU synonym for
.BR FNM_PATHNAME .
.TP
.B FNM_LEADING_DIR
If this flag (a GNU extension) is set, the pattern is considered to be
matched if it matches an initial segment of
.I string
which is followed by a slash.
This flag is mainly for the internal
use of glibc and is implemented only in certain cases.
.TP
.B FNM_CASEFOLD
.TQ
.BR FNM_IGNORECASE " (same as " FNM_CASEFOLD )
If this flag is set, the pattern is matched
case-insensitively.
.TP
.B FNM_EXTMATCH
If this flag (a GNU extension) is set, extended patterns are
supported, as introduced by \&'ksh' and now supported by other shells.
The extended format is as follows, with
.I pattern\-list
being a \&'|' separated list of patterns.
.TP
.RI \&'?( pattern\-list )'
The pattern matches if zero or one occurrences of any of the
patterns in the
.I pattern\-list
match the input
.IR string .
.TP
.RI \&'*( pattern\-list )'
The pattern matches if zero or more occurrences of any of the
patterns in the
.I pattern\-list
match the input
.IR string .
.TP
.RI \&'+( pattern\-list )'
The pattern matches if one or more occurrences of any of the
patterns in the
.I pattern\-list
match the input
.IR string .
.TP
.RI \&'@( pattern\-list )'
The pattern matches if exactly one occurrence of any of the
patterns in the
.I pattern\-list
match the input
.IR string .
.TP
.RI \&'!( pattern\-list )'
The pattern matches if the input
.I string
cannot be matched with
any of the patterns in the
.IR pattern\-list .
.SH RETURN VALUE
Zero if
.I string
matches
.IR pattern ,
.B FNM_NOMATCH
if there is no match or another nonzero value if there is an error.
.SH ATTRIBUTES
For an explanation of the terms used in this section, see
.BR attributes (7).
.TS
allbox;
lbx lb lb
l l l.
Interface	Attribute	Value
T{
.na
.nh
.BR fnmatch ()
T}	Thread safety	MT-Safe env locale
.TE
.SH STANDARDS
.TP
.BR fnmatch ()
POSIX.1-2008.
.TP
.B FNM_CASEFOLD
.TQ
.B FNM_IGNORECASE
POSIX.1-2024.
.TP
.B FNM_FILE_NAME
.TQ
.B FNM_LEADING_DIR
GNU.
.SH HISTORY
.TP
.BR fnmatch ()
POSIX.1-2001, POSIX.2.
.TP
.B FNM_CASEFOLD
has been available on many systems even before POSIX.1-2024.
.SH SEE ALSO
.BR sh (1),
.BR glob (3),
.BR scandir (3),
.BR wordexp (3),
.BR glob (7)


man-pages/Changes
==================== Changes in man-pages-6.16 ====================

Released: 2025-10-29, Aldaya


New and rewritten pages
-----------------------

man2/
	fsconfig.2
	fsmount.2
	fsopen.2
	fspick.2
	move_mount.2
	open_tree.2

man2const/
	PR_FUTEX_HASH_GET_SLOTS.2const
	PR_FUTEX_HASH_SET_SLOTS.2const

man2type/
	mount_attr.2type			(split from mount_setattr(2))

man3/
	_Fork.3
	countof.3
	memeq.3
	streq.3


Newly documented interfaces in existing pages
---------------------------------------------

man2/
	mount_setattr.2
		glibc wrapper

	pidfd_open.2
		PIDFD_THREAD

	pidfd_send_signal.2
		PIDFD_SIGNAL_THREAD
		PIDFD_SIGNAL_THREAD_GROUP
		PIDFD_SIGNAL_PROCESS_GROUP

	quotactl.2
		quotactl_fd(2)

	readv.2
		RWF_DONTCACHE

	sigaction.2
		SI_ASYNCNL
		FPE_FLTUNK
		FPE_CONDTRAP
		SEGV_ACCADI
		SEGV_ADIDERR
		SEGV_ADIPERR
		SEGV_MTEAERR
		SEGV_MTESERR
		SEGV_CPERR
		TRAP_UNK
		TRAP_PERF
		SYS_USER_DISPATCH
		EMT_TAGOVF

man3/
	abs.3
		umaxabs(3)

	opendir.3
		ENAMETOOLONG

man7/
	operator.7
		(type){}
		alignof
		_Countof

	rtnetlink.7
		RTAX_UNSPEC
		RTAX_LOCK
		RTAX_MTU
		RTAX_WINDOW
		RTAX_RTT
		RTAX_RTTVAR
		RTAX_SSTHRESH
		RTAX_CWND
		RTAX_ADVMSS
		RTAX_REORDERING
		RTAX_HOPLIMIT
		RTAX_INITCWND
		RTAX_FEATURES
		RTAX_RTO_MIN
		RTAX_INITRWND
		RTAX_QUICKACK
		RTAX_CC_ALGO
		RTAX_FASTOPEN_NO_COOKIE

	sched.7
		CONFIG_PREEMPT
		CONFIG_PREEMPT_LAZY

	tcp.7
		TCP_SAVE_SYN
		TCP_SAVED_SYN


New and changed links
---------------------

man2/
	open_tree_attr.2			(open_tree(2))
	quotactl_fd.2				(quotactl(2))
man3/
	umaxabs.3				(abs(3))


Global changes
--------------

-  man/
   -  Update documentation for POSIX.1-2024 (still more work to do).
   -  SYNOPSIS: Use GNU forward-declarations of parameters for sizes of
      array parameters.
   -  Improve style of source code for tbl(1) tables.
   -  man2/: Use the common name 'fd' instead of 'fildes'.
   -  Use semantic newlines.
   -  EXAMPLES
      -  Use err(3) and errc(3bsd) instead of similar macros.
      -  Fix includes (thanks to iwyu(1)).
      -  Use NITEMS() consistently.
      -  Don't name unused parameters.
-  share/mk/
   -  Add support for the chapter man3attr.
   -  help: Document the usual targets.
   -  lint-man-blank: Add target to lint about blank lines.
   -  lint-man-ws: Diagnose spurious use of white space.
   -  lint-man-quote: Diagnose an unmatched quote.
   -  lint-man-poems: Diagnose (lack of) semantic newlines.
   -  lint-man-dash: Diagnose unescaped dashes
   -  Merge handling of man(7) and mdoc(7) pages.
   -  lint, build, check: Move exceptions to separate files.
   -  Remove unused variables.
   -  Unify local variable names.
   -  Silence diagnostics about unnamed parameters.
   -  configure/build-depends/:
      -  Add awk/awk.
      -  Add pcre2-utils/pcre2grep.


Changes to individual files
---------------------------

-  src/bin/
   -  sortman(1): Fix order of intro(*) pages for subsections.
   -  diffman-git(1): Show sections as hunk context.

The manual pages and other files in the repository have been improved
beyond what this changelog covers.  To learn more about changes applied
to individual pages, or the authors of changes, use git(1).


src/mandb/lib/fnmatch.c
/* Copyright (C) 1991, 1992, 1993 Free Software Foundation, Inc.

This library is free software; you can redistribute it and/or
modify it under the terms of the GNU Library General Public License as
published by the Free Software Foundation; either version 2 of the
License, or (at your option) any later version.

This library is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
Library General Public License for more details.

You should have received a copy of the GNU Library General Public
License along with this library; see the file COPYING.LIB.  If
not, write to the Free Software Foundation, Inc., 675 Mass Ave,
Cambridge, MA 02139, USA.  */

#ifdef HAVE_CONFIG_H
#include "config.h"
#endif

#include <errno.h>
#include "fnmatch.h"
#include <ctype.h>


/* Comment out all this code if we are using the GNU C Library, and are not
   actually compiling the library itself.  This code is part of the GNU C
   Library, but also included in many other GNU distributions.  Compiling
   and linking in this code is a waste when using the GNU C library
   (especially if it is a shared library).  Rather than having every GNU
   program understand `configure --with-gnu-libc' and omit the object files,
   it is simpler to just do this in the source for each such file.  */

#if defined (_LIBC) || !defined (__GNU_LIBRARY__)


#if !defined(__GNU_LIBRARY__) && !defined(STDC_HEADERS)
extern int errno;
#endif

/* Match STRING against the filename pattern PATTERN, returning zero if
   it matches, nonzero if not.  */
int
fnmatch (pattern, string, flags)
     const char *pattern;
     const char *string;
     int flags;
{
  register const char *p = pattern, *n = string;
  register char c;

/* Note that this evalutes C many times.  */
#define FOLD(c)	((flags & FNM_CASEFOLD) && isupper (c) ? tolower (c) : (c))

  while ((c = *p++) != '\0')
    {
      c = FOLD (c);

      switch (c)
	{
	case '?':
	  if (*n == '\0')
	    return FNM_NOMATCH;
	  else if ((flags & FNM_FILE_NAME) && *n == '/')
	    return FNM_NOMATCH;
	  else if ((flags & FNM_PERIOD) && *n == '.' &&
		   (n == string || ((flags & FNM_FILE_NAME) && n[-1] == '/')))
	    return FNM_NOMATCH;
	  break;

	case '\\':
	  if (!(flags & FNM_NOESCAPE))
	    {
	      c = *p++;
	      c = FOLD (c);
	    }
	  if (FOLD (*n) != c)
	    return FNM_NOMATCH;
	  break;

	case '*':
	  if ((flags & FNM_PERIOD) && *n == '.' &&
	      (n == string || ((flags & FNM_FILE_NAME) && n[-1] == '/')))
	    return FNM_NOMATCH;

	  for (c = *p++; c == '?' || c == '*'; c = *p++, ++n)
	    if (((flags & FNM_FILE_NAME) && *n == '/') ||
		(c == '?' && *n == '\0'))
	      return FNM_NOMATCH;

	  if (c == '\0')
	    return 0;

	  {
	    char c1 = (!(flags & FNM_NOESCAPE) && c == '\\') ? *p : c;
	    c1 = FOLD (c1);
	    for (--p; *n != '\0'; ++n)
	      if ((c == '[' || FOLD (*n) == c1) &&
		  fnmatch (p, n, flags & ~FNM_PERIOD) == 0)
		return 0;
	    return FNM_NOMATCH;
	  }

	case '[':
	  {
	    /* Nonzero if the sense of the character class is inverted.  */
	    register int not;

	    if (*n == '\0')
	      return FNM_NOMATCH;

	    if ((flags & FNM_PERIOD) && *n == '.' &&
		(n == string || ((flags & FNM_FILE_NAME) && n[-1] == '/')))
	      return FNM_NOMATCH;

	    not = (*p == '!' || *p == '^');
	    if (not)
	      ++p;

	    c = *p++;
	    for (;;)
	      {
		register char cstart = c, cend = c;

		if (!(flags & FNM_NOESCAPE) && c == '\\')
		  cstart = cend = *p++;

		cstart = cend = FOLD (cstart);

		if (c == '\0')
		  /* [ (unterminated) loses.  */
		  return FNM_NOMATCH;

		c = *p++;
		c = FOLD (c);

		if ((flags & FNM_FILE_NAME) && c == '/')
		  /* [/] can never match.  */
		  return FNM_NOMATCH;

		if (c == '-' && *p != ']')
		  {
		    cend = *p++;
		    if (!(flags & FNM_NOESCAPE) && cend == '\\')
		      cend = *p++;
		    if (cend == '\0')
		      return FNM_NOMATCH;
		    cend = FOLD (cend);

		    c = *p++;
		  }

		if (FOLD (*n) >= cstart && FOLD (*n) <= cend)
		  goto matched;

		if (c == ']')
		  break;
	      }
	    if (!not)
	      return FNM_NOMATCH;
	    break;

	  matched:;
	    /* Skip the rest of the [...] that already matched.  */
	    while (c != ']')
	      {
		if (c == '\0')
		  /* [... (unterminated) loses.  */
		  return FNM_NOMATCH;

		c = *p++;
		if (!(flags & FNM_NOESCAPE) && c == '\\')
		  /* XXX 1003.2d11 is unclear if this is right.  */
		  ++p;
	      }
	    if (not)
	      return FNM_NOMATCH;
	  }
	  break;

	default:
	  if (c != FOLD (*n))
	    return FNM_NOMATCH;
	}

      ++n;
    }

  if (*n == '\0')
    return 0;

  if ((flags & FNM_LEADING_DIR) && *n == '/')
    /* The FNM_LEADING_DIR flag says that "foo*" matches "foobar/frobozz".  */
    return 0;

  return FNM_NOMATCH;
}

#endif	/* _LIBC or not __GNU_LIBRARY__.  */
